src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+            p.eat();
+        })
+    }).collect();
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+
+    for h in handles {
+        h.join().ok().expect("Couldn't join a thread.");
+    }
+}
+```
+
+All we’ve done is change the loop in `main()`, and added a third one! Here’s the
+first change:
+
+```rust,ignore
+let handles: Vec<_> = philosophers.into_iter().map(|p| {
+    thread::spawn(move || {
+        p.eat();
+    })
+}).collect();
+```
+
+While this is only five lines, they’re a dense four. Let’s break it down.
+
+```rust,ignore
+let handles: Vec<_> = 
+```
+
+We introduce a new binding, called `handles`. We’ve given it this name because
+we are going to make some new threads, and that will return some handles to those
+threads that let us control their operation. We need to explicitly annotate
+the type here, though, due to an issue we’ll talk about later. The `_` is
+a type placeholder. We’re saying “`handles` has is a vector of something, but
+you can figure out what that something is, Rust.”
+
+```rust,ignore
+philosophers.into_iter().map(|p| {
+```
+
+We take our list of philosophers and call `into_iter()` on it. This creates an
+iterator that takes ownership of each philosopher. We need to do this to pass
+them to our threads. We take that iterator and call `map` on it, which takes a
+closure as an argument and calls that closure on each element in turn.
+
+```rust,ignore
+    thread::spawn(move || {
+        p.eat();
+    })
+```
+
+Here’s where the concurrency happens. The `thread::spawn` function takes a closure
+as an argument and executes that closure in a new thread. This closure needs
+an extra annotation, `move`, to indicate that the closure is going to take
+ownership of the values it’s capturing. Primarily, the `p` variable of the
+`map` function.
+
+Inside the thread, all we do is call `eat()` on `p`.
+
+```rust,ignore
+}).collect();
+```
+
+Finally, we take the result of all those `map` calls and collect them up.
+`collect()` will make them into a collection of some kind, which is why we
+needed to annotate the return type: we want a `Vec<T>`. The elements are the
+return values of the `thread::spawn` calls, which are handles to those threads.
+Whew!
+
+
+```rust,ignore
+for h in handles {
+    h.join().ok().expect("Couldn't join a thread.");
+}
+```
+
+At the end of `main()`, we loop through the handles and call `join()` on them,
+which blocks execution until the thread has completed execution. This ensures
+that the threads complete their work before the program exits.
+
+If you run this program, you’ll see that the philosophers eat out of order!
+We have mult-threading!
+
+```text
+Gilles Deleuze is eating.
+Gilles Deleuze is done eating.
+Friedrich Nietzsche is eating.
+Friedrich Nietzsche is done eating.
+Michel Foucault is eating.
+Baruch Spinoza is eating.
+Baruch Spinoza is done eating.
+Karl Marx is eating.
+Karl Marx is done eating.
+Michel Foucault is done eating.
```

Comment:
I wonder if perhaps the wrong binary was run? I get this for example:

```
Gilles Deleuze is eating.
Baruch Spinoza is eating.
Michel Foucault is eating.
Friedrich Nietzsche is eating.
Karl Marx is eating.
Baruch Spinoza is done eating.
Gilles Deleuze is done eating.
Michel Foucault is done eating.
Karl Marx is done eating.
Friedrich Nietzsche is done eating.
```

There's also quite a noticeable pause (as expected) between when everyone starts eating and when they finish.